File: Solved/Baste Dar Jadval/Baste Dar Jadval.py
# بسته در جدول
# ID : 106797
# https://quera.org/problemset/106797/
# https://b2n.ir/m78821

def isPrime(num):
    if num == 1:
        return False

    i = 2
    while i * i <= num:
        if num % i == 0:
            return False

        i += 1

    return True


def findCenter(num):
    return (num / 2) + 0.5


def findOpposite(length, num):
    center = findCenter(length)

    result = 0
    if length % 2 == 0:
        if num < center:
            result = int(center + (center - num))

        elif num > center:
            result = int(center - (num - center))

    else:
        if num == int(center):
            result = num

        elif num < center:
            result = int(center + (center - num))

        elif num > center:
            result = int(center - (num - center))

    return result


line = [int(i) for i in input().split(' ')]

row, column, count = line

keep = []

for i in range(row):
    inputer = input().split(' ')
    keep.append([])
    for j in range(column):
        keep[i].append(int(inputer[j]))

current = keep[0][0]
current_row = 0
current_column = 0

for i in range(count):
    if isPrime(current):
        current_row = findOpposite(row, current_row + 1) - 1
        current_column = findOpposite(column, current_column + 1) - 1
        current = keep[current_row][current_column]

    else:
        if current % 4 == 0:
            if current_column == column - 1:
                current_column = 0
            else:
                current_column += 1
            current = keep[current_row][current_column]

        elif current % 4 == 1:
            if current_row == row - 1:
                current_row = 0
            else:
                current_row += 1

            current = keep[current_row][current_column]

        elif current % 4 == 2:
            if current_column == 0:
                current_column = column - 1
            else:
                current_column -= 1

            current = keep[current_row][current_column]

        elif current % 4 == 3:
            if current_row == 0:
                current_row = row - 1
            else:
                current_row -= 1

            current = keep[current_row][current_column]

print(f'{current_row + 1} {current_column + 1}')

File: Solved/Baste Dar Jadval/BasteDarJadval.java
import java.util.Scanner;

/*
Question title = baste dar jadval - بسته در جدول
https://quera.ir/problemset/106797/
 */
public class BasteDarJadval {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine();

        int row = Integer.parseInt(line.split("\\s")[0]);
        int column = Integer.parseInt(line.split("\\s")[1]);
        int count = Integer.parseInt(line.split("\\s")[2]);

        int[][] list = new int[row][column];

        //getting table values from user
        for (int i = 0; i <= row - 1; i++) {
            String input = sc.nextLine();
            for (int j = 0; j <= column - 1; j++) {
                list[i][j] = Integer.parseInt(input.split("\\s")[j]);
            }
        }
        //current value
        int current = list[0][0];
        //current row index
        int currentRow = 0;
        //current column index
        int currentColumn = 0;

        for (int i = 0; i <= count - 1; i++) {
            //if current number is prime , than go to opposite row and column
            if (isPrime(current)) {
                currentRow = findOpposite(row, currentRow + 1) - 1;
                currentColumn = findOpposite(column, currentColumn + 1) - 1;
                current = list[currentRow][currentColumn];

            } else {
                switch (current % 4) {
                    case 0:
                        if (currentColumn == column - 1) {
                            currentColumn = 0;
                        } else {
                            currentColumn++;
                        }
                        current = list[currentRow][currentColumn];
                        break;
                    case 1:
                        if (currentRow == row - 1) {
                            currentRow = 0;
                        } else {
                            currentRow++;
                        }
                        current = list[currentRow][currentColumn];

                        break;
                    case 2:
                        if (currentColumn == 0) {
                            currentColumn = column - 1;
                        } else {
                            currentColumn--;
                        }
                        current = list[currentRow][currentColumn];

                        break;
                    case 3:
                        if (currentRow == 0) {
                            currentRow = row - 1;
                        } else {
                            currentRow--;
                        }
                        current = list[currentRow][currentColumn];
                        break;
                }
            }
        }
        System.out.println(currentRow + 1 + " " + (currentColumn + 1));
    }

    //check if a number is prime or not
    private static boolean isPrime(int num) {
        boolean isPrime = true;
        for (int i = 2; i <= num / 2 + 1; i++) {
            if (num % i == 0) {
                isPrime = false;
                break;
            }
        }
        return isPrime;
    }

    //returns middle cell in given table length
    private static double findCenter(int num) {

        return (num / 2.0) + 0.5;
    }

    //returns opposite cell in given row or column
    private static int findOpposite(int length, int num) {
        double center = findCenter(length);
        //even
        int result = 0;
        if (length % 2 == 0) {
            if (num < center) {
                result = (int) (center + (center - num));
            }
            if (num > center) {
                result = (int) (center - (num - center));
            }
        }
        //odd
        else {
            if (num == (int) center) {
                result = num;
            }
            if (num < center) {
                result = (int) (center + (center - num));
            }
            if (num > center) {
                result = (int) (center - (num - center));
            }
        }
        return result;
    }
}